### PR TITLE
Allow multiple lifters to contribute to the lifting process (spotters)

### PR DESCRIPTION
**Pull request by Nilocunger**

### CODE REVIEW

**Comment by ltfish** [comment]
Conflicts!

**Comment by zardus** [comment]
What's the road map for merging this? Just fixing tests, or is this still very WIP?

I ask because I'm designing my course for next semester, and I'd like one assignment to be the creation of a lifter for a previously-unsupported arch, and I'd like to figure out what version of API and so forth to have my victims^Wstudents target.

**Comment by Nilocunger** [comment]
Almost ready. The branch passes all but a couple tests.

File: pyvex/block.py [hunk 1]
```diff
@@ -69,14 +71,64 @@ def __init__(self, data, mem_addr, arch, max_inst=None, max_bytes=None, bytes_of
         self._size = None
 
         if data is not None:
-            lift(self, data, max_bytes, max_inst, bytes_offset, opt_level, traceflags)
+            lift(self, arch, mem_addr, data, max_bytes, max_inst, bytes_offset, opt_level, traceflags)
+
+    @staticmethod
+    def emptyBlock(arch, addr, statements=None, nxt=None, tyenv=None, jumpkind=None, direct_next=None, size=None):
```
Comment (rhelmot) [inline pyvex/block.py]: snake_case forever and always

File: pyvex/block.py [hunk 2]
```diff
@@ -69,14 +71,64 @@ def __init__(self, data, mem_addr, arch, max_inst=None, max_bytes=None, bytes_of
...
+        self.next = extendwith.next
+        self.jumpkind = extendwith.jumpkind
 
     def pp(self):
         """
         Pretty-print the IRSB to stdout.
         """
         print(self._pp_str())
 
+    def __str__(self):
```
Comment (rhelmot) [inline pyvex/block.py]: this'll be obnoxious unless you add a `__repr__` which prints a single-line description

File: pyvex/lift/libvex.py [hunk 3]
```diff
@@ -3,49 +3,110 @@
...
+from . import Lifter, register, LiftingException
 from .. import pvc, ffi
+from ..enums import default_vex_archinfo, vex_endness_from_string
 
 _libvex_lock = threading.Lock()
 
 SUPPORTED = {'X86', 'AMD64', 'MIPS32', 'MIPS64', 'ARM', 'ARMEL', 'ARMHF', 'AARCH64', 'PPC32', 'PPC64'}
 
+VEX_MAX_INSTRUCTIONS = 99
+VEX_MAX_BYTES = 400
```
Comment (rhelmot) [inline pyvex/lift/libvex.py]: this 400 value is actually a lie, it's a lot bigger! at least 1000, I can't recall

Comment (rhelmot) [reply pyvex/lift/libvex.py]: libvex.h says 5000

File: pyvex/block.py [hunk 4]
```diff
@@ -69,14 +71,67 @@ def __init__(self, data, mem_addr, arch, max_inst=None, max_bytes=None, bytes_of
...
+        invalid_vals = (0xffffffff, -1)
+
+        def convert_tmp(tmp):
+            if tmp not in conversion_dict:
+                tmp_type = extendwith.tyenv.lookup(tmp)
+                conversion_dict[tmp] = self.tyenv.add(tmp_type)
+            return conversion_dict[tmp]
+
+        def convert_expr(expr):
+            if isinstance(expr, RdTmp):
```
Comment (subwire) [inline pyvex/block.py]: A super brief comment with what these are for (so we can maintain them later) would help a lot.

File: pyvex/lift/__init__.py [hunk 5]
```diff
@@ -9,62 +17,51 @@ class Lifter(object):
...
+             max_inst=None,
+             opt_level=1,
+             traceflags=None,
+             allow_lookback=None):
+        """
+        Create an irsb from data
+
+        All of the lifters will be used in the order that they are registered
+        whenever an instruction cannot be decoded until one is able to decode it
+        """
```
Comment (subwire) [inline pyvex/lift/__init__.py]: The parameters to these were always a bit mysterious.  I know some of these get defined a few other places, but seeing as this is a base class for a few other important things (and for people to write more Lifters sans-Gymrat), perhaps we should lay down a real docstring here?

File: pyvex/lift/__init__.py [hunk 6]
```diff
@@ -9,62 +17,51 @@ class Lifter(object):
...
-    if not max_bytes:
-        max_bytes = len(data)
-
-    if max_bytes == 0:
-        raise PyVEXError("No bytes provided")
-
-    errors = []
-
+def lift(irsb, arch, addr, data, max_bytes=None, max_inst=None, bytes_offset=None, opt_level=1, traceflags=False):
+    final_irsb = IRSB.empty_block(arch, addr)
```
Comment (subwire) [inline pyvex/lift/__init__.py]: Probably here too

File: pyvex/lift/__init__.py [hunk 7]
```diff
@@ -75,51 +72,67 @@ def lift(irsb, data, max_bytes, max_inst, bytes_offset, opt_level, traceflags):
...
+
+    for postprocessor in postprocessors:
+        try:
+            postprocessor(final_irsb).postprocess()
+        except LiftingException:
+            continue
+    irsb._from_py(final_irsb)
 
 def register(lifter):
-    if lifter.lift is not Lifter.lift:
```
Comment (subwire) [inline pyvex/lift/__init__.py]: Definitely here, as we changed how this works.

**Pull request merged**